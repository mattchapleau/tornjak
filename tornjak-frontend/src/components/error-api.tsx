import { toast, ToastContent, ToastOptions } from "react-toastify"
import { ToastNotification, ToastNotificationProps } from "carbon-components-react"

interface NotificationProps extends Omit<ToastNotificationProps, "title"> {
    title?: string
}

const defualtProps: ToastNotificationProps = {title: "Notification", kind: "error"}
const defaultOptions: ToastContent = {autoClose: false, closeButton: false, role: "alert"}

export const showToast = (props?: NotificationProps, options?: ToastOptions): void => {
    const newProps = {...defualtProps, ...props}
    toast(<ToastNotification {...newProps} />, {...defaultOptions, ...options})
}

type Response = {response: {data: string, status: number}}

const defaultResponseProps = (res: Response): NotificationProps => {
    if (res.response === undefined) {
        return {caption: "Could not connect to backend", title: "Network Error"}
    }
    return {caption: res.response.data, title: "Error " + String(res.response.status)}
}

export const showResponseToast = (res: Response, props?: NotificationProps, options?: ToastOptions): void => {
    showToast({...defaultResponseProps(res), ...props}, options)
}