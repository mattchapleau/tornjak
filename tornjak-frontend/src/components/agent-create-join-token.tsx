import { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import GetApiServerUri from './helpers';
import IsManager from './is_manager';
import { serverSelectedFunc } from 'redux/actions';
import { RootState } from 'redux/reducers';
import { ToastContainer } from "react-toastify"
import { showResponseToast, showToast } from './error-api';

type CreateJoinTokenProp = {
  globalServerSelected: string,
}

type CreateJoinTokenState = {
  name: string,
  ttl: number | string,
  token: string,
  spiffeId: string,
  trustDomain: string,
  path: string,
  message: string,
  selectedServer: string,
}

class CreateJoinToken extends Component<CreateJoinTokenProp, CreateJoinTokenState> {
  constructor(props: CreateJoinTokenProp) {
    super(props);
    this.onChangeName = this.onChangeName.bind(this);
    this.onChangeTtl = this.onChangeTtl.bind(this);
    this.onChangeToken = this.onChangeToken.bind(this);
    this.onChangeSpiffeId = this.onChangeSpiffeId.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.state = {
      name: "",
      ttl: 500,
      token: "",
      spiffeId: "",
      trustDomain: "",
      path: "",
      message: "",
      selectedServer: "",
    }
  }

  componentDidMount() {
    if (IsManager) {
      if (this.props.globalServerSelected !== "") {
        this.setState({ selectedServer: this.props.globalServerSelected });
      }
    } else {
      // agent doesnt need to do anything
      this.setState({})
    }
  }

  componentDidUpdate(prevProps: CreateJoinTokenProp) {
    if (IsManager) {
      if (prevProps.globalServerSelected !== this.props.globalServerSelected) {
        this.setState({ selectedServer: this.props.globalServerSelected });
      }
    }
  }

  onChangeName(e: { target: { value: string; }; }): void {
    this.setState({
      name: e.target.value
    });
  }

  onChangeTtl(e: { target: { value: string; }; }): void {
    this.setState({
      ttl: e.target.value ? Number(e.target.value) : String(e.target.value)
    });
  }

  onChangeToken(e: { target: { value: string; }; }): void {
    this.setState({
      token: e.target.value
    });
  }

  parseSpiffeId(sid: string): [boolean, string, string] {
    var prefix = "spiffe://";
    if (sid.startsWith(prefix)) {
      var sub = sid.substr(prefix.length)
      var sp = sub.indexOf("/")
      if (sp > 0 && sp !== sub.length - 1) {
        var trustDomain = sub.substr(0, sp);
        var path = sub.substr(sp);
        return [true, trustDomain, path];
      }
    }
    return [false, "", ""];
  }

  onChangeSpiffeId(e: { target: { value: string; }; }): void {
    this.setState({
      spiffeId: e.target.value
    })
  }

  getApiTokenEndpoint(): string {
    if (!IsManager) {
      return GetApiServerUri('/api/agent/createjointoken')
    } else if (IsManager && this.state.selectedServer !== "") {
      return GetApiServerUri('/manager-api/agent/createjointoken') + "/" + this.state.selectedServer
    } else {
      this.setState({ message: "Error: No server selected" })
      return ""
    }


  }
  onSubmit(e: { preventDefault: () => void; }): void {
    e.preventDefault()

    if (this.state.ttl === "") {
      showToast({caption: "The TTL cannot be empty."})
      return
    }

    if (this.state.ttl <= 0) {
      showToast({caption: "The TTL must be positive."})
      return
    }

    const cjtData = {
      ttl: this.state.ttl, 
      trust_domain: "", 
      path: "", 
      token: this.state.token
    }

    if (this.state.spiffeId) {
      const [isIdValid, trustDomain, path] = this.parseSpiffeId(this.state.spiffeId)
      
      if (!isIdValid) {
        showToast({caption: "The spiffe id is invalid."})
        return
      }

      cjtData.trust_domain = trustDomain
      cjtData.path = path
    }

    let endpoint = this.getApiTokenEndpoint()

    if (!endpoint) {
      return
    }

    axios.post(endpoint, cjtData)
      .then(res => this.setState({ message: "Request:" + JSON.stringify(cjtData, null, ' ') + "\n\nSuccess:" + JSON.stringify(res.data, null, ' ') }))
      .catch(err => showResponseToast(err))
  }

  render() {
    return (
      <div data-test="agent-create-token">
        <h3>Create New Agent Join Token</h3>
        <form onSubmit={this.onSubmit}>
          <div className="alert-primary" role="alert">
            <pre>
              {this.state.message}
            </pre>
          </div>
          {IsManager}
          <br /><br />

          <div className="form-group">
            <label>Time to live (TTL): </label>
            <input type="number"
              className="form-control"
              value={this.state.ttl}
              onChange={this.onChangeTtl}
            />
          </div>


          <div className="form-group">
            <label>Token (optional): i.e. 1adcc067-18e5-4d6f-be97-aa74b5ba5c28</label>
            <input type="text"
              className="form-control"
              value={this.state.token}
              onChange={this.onChangeToken}
            />
          </div>

          <div className="form-group">
            <label>SPIFFE ID (optional): i.e. spiffe://example.org/sample/spiffe/id</label>
            <input type="text"
              className="form-control"
              value={this.state.spiffeId}
              onChange={this.onChangeSpiffeId}
            />
          </div>
          <div className="form-group">
            <input type="submit" value="Create Join Token" className="btn btn-primary" />
          </div>
        </form>
        <ToastContainer
          className="carbon-toast"
          containerId="notifications"
          draggable={false}
        />
      </div>
    )
  }
}

const mapStateToProps = (state: RootState) => ({
  globalServerSelected: state.servers.globalServerSelected,
})

// Note: Needed for UI testing - will be removed after
// CreateJoinToken.propTypes = {
//   serverSelectedFunc: PropTypes.func,
//   globalServerSelected: PropTypes.string,
// };

export default connect(
  mapStateToProps,
  { serverSelectedFunc }
)(CreateJoinToken)

export { CreateJoinToken };