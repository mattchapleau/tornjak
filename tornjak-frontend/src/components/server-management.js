import { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios'
import GetApiServerUri from './helpers';
import IsManager from './is_manager';
import {
  serversListUpdateFunc
} from 'redux/actions';
import { showResponseToast } from './error-api';

const Server = props => (
  <tr>
    <td>{props.server.name}</td>
    <td>{props.server.address}</td>
    <td>{(props.server.mtls && "mTLS") || (props.server.tls && "TLS") || "None"}</td>
  </tr>
)

class ServerManagement extends Component {
  constructor(props) {
    super(props);
    this.state = {
        formServerName: "",
        formServerAddress: "",
        formTLS: false,
        formMTLS: false,
        formCAData: null,
        formCertData: null,
        formKeyData: null,
        CAFileText: "",
        certFileText: "",
        keyFileText: "",
    };
    this.onCertFileChange = this.onCertFileChange.bind(this);
    this.onCAFileChange = this.onCAFileChange.bind(this);
    this.onKeyFileChange = this.onKeyFileChange.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentDidMount() {
      this.refreshServerState()
  }

  refreshServerState () {
    axios.get(GetApiServerUri("/manager-api/server/list"), { crossdomain: true })
      .then(response => {
          console.log(response.data);
          this.props.serversListUpdateFunc(response.data["servers"]);
      })
      .catch((error) => showResponseToast(error, {caption: "Could not refresh server state."}))
  }

  serverList() {
    if (typeof this.props.globalServersList !== 'undefined') {
        return this.props.globalServersList.map(s => {
          return <Server key={s.name} 
                    server={s} />;
        })
    } else {
        return ""
    }
  }

  handleInputChange(e) {
    const target = e.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.setState({
      [name]: value
    });
  }


  onCAFileChange = event => {
    // Update the state
    const reader = new FileReader();
    reader.onload = e => {
      this.setState({ 
          formCAData: (new Buffer(e.target.result)).toString("base64"),
          CAFileText: "CA file load success",
      });
    }
    reader.readAsText(event.target.files[0])
  };
  

  onCertFileChange = event => {
    // Update the state
    const reader = new FileReader();
    reader.onload = e => {
      this.setState({ 
          formCertData: (new Buffer(e.target.result)).toString("base64"),
          certFileText: "cert file load success",
      });
    }
    reader.readAsText(event.target.files[0])
  };

  onKeyFileChange = event => {
    // Update the state
    const reader = new FileReader();
    reader.onload = e => {
      this.setState({ 
          formKeyData: (new Buffer(e.target.result)).toString("base64"),
          keyFileText: "key file load success",
      });
    }
    reader.readAsText(event.target.files[0])
  };


  onSubmit(e) {
    e.preventDefault();

    console.log("onSubmit");
    var cjtData = {
        "name": this.state.formServerName,
        "address": this.state.formServerAddress,
        "tls": this.state.formTLS,
        "mtls": this.state.formMTLS,
        "ca": this.state.formCAData,
        "cert": this.state.formCertData,
        "key": this.state.formKeyData,
    };
    axios.post(GetApiServerUri('/manager-api/server/register'), cjtData)
      .then(res => {
          this.setState({ message: "Requst:" + JSON.stringify(cjtData,null,  ' ')+ "\n\nSuccess:" + JSON.stringify(res.data, null, ' ')});
          this.refreshServerState();
      }
      )
      .catch(err => this.setState({ message: "ERROR:" + err + (typeof (err.response) !== "undefined" ? err.response.data : "")}))

  }


  render() {
    if (!IsManager) {
        return <h1>Only manager deployments have use of this page</h1>
    }

    const tlsFormOptions = (
          <div> CA File (for (m)TLS):
          <input name="CAfile" type="file" onChange={this.onCAFileChange} />
            {this.state.CAFileText}
          </div>
    )

    const mtlsFormOptions = (
        <div>
          <div> 
           Cert File (for mTLS):
          <input name="certfile" type="file" onChange={this.onCertFileChange} />
            {this.state.certFileText}
          </div>
          <div> Key File (for mTLS):
          <input name="keyfile" type="file" onChange={this.onKeyFileChange} />
            {this.state.keyFileText}
          </div>
        </div>
    )

    return (
      <div>
        <h3> Register New Server </h3>
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label>Server Name (Unique)</label>
            <input type="text"
              name="formServerName"
              required
              className="form-control"
              value={this.state.formServerName}
              onChange={this.handleInputChange}
            />
          </div>


          <div className="form-group">
            <label>Address (i.e. http://localhost:5000/) </label>
            <input type="text"
              name="formServerAddress"
              required
              className="form-control"
              value={this.state.formServerAddress}
              onChange={this.handleInputChange}
            />
          </div>

          <div name="tls-checkbox-input">
          <input
            name="formTLS"
            type="checkbox"
            checked={this.state.formTLS}
            onChange={this.handleInputChange}
            />
              TLS Enabled
          </div>
          {this.state.formTLS && tlsFormOptions}

          <div name="mtls-checkbox-input">
          <input
            name="formMTLS"
            type="checkbox"
            checked={this.state.formMTLS}
            onChange={this.handleInputChange}
            />
            mTLS Enabled
          </div>
          {this.state.formMTLS && mtlsFormOptions}

          <div className="form-group">
            <br></br>
          <input type="submit" value="Register Server" className="btn btn-primary" />
          </div>
        </form>

        <div className="alert-primary" role="alert">
        <pre>
           {this.state.message}
        </pre>
        </div>


        <h3>Server List</h3>
        <table className="table" style={{width : "100%"}}>
          <thead className="thead-light">
            <tr>
              <th>Server Name</th>
              <th>Address</th>
              <th>TLS?</th>
            </tr>
          </thead>
          <tbody>
            {this.serverList()}
          </tbody>
        </table>

      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  globalServersList: state.servers.globalServersList
})

export default connect(
  mapStateToProps,
  { serversListUpdateFunc }
)(ServerManagement)
