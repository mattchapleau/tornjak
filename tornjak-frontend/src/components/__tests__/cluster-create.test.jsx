import { ClusterCreate } from "../../components/cluster-create"
import { shallow, configure } from "enzyme"
import Adapter from "@wojtekmaj/enzyme-adapter-react-17"

configure({ adapter: new Adapter() })

const props = {
    serverInfoUpdateFunc: jest.fn(), 
    agentsListUpdateFunc: jest.fn(), 
    tornjakMessageFunc: jest.fn(), 
    tornjakServerInfoUpdateFunc: jest.fn(), 
    clusterTypeInfoFunc: jest.fn(), 
    agentsList: [{}], 
    clusterTypeList: [], 
    globalServerSelected: "Test String", 
    globalErrorMessage: "Test String", 
    globalTornjakServerInfo: {}
}

describe("Cluster Create Component", () => {
    const wrapper = shallow(<ClusterCreate {...props} />)

    it("Should Render", () => {
        expect(wrapper).toMatchSnapshot()
    })

    it("Should Create Cluster", () => {
        const instance = wrapper.instance()


        instance.onChangeClusterName({target: {value: "Name"}})
        instance.onChangeManualClusterType({target: {value: "Type"}})
        instance.onChangeClusterDomainName({target: {value: "Domain"}})
        instance.onChangeClusterManagedBy({target: {value: "Managed By"}})

        wrapper.find("form").simulate("submit")

        expect(wrapper.state("clusterName")).toBe("Name")
        expect(wrapper.state("clusterType")).toBe("Type")
        expect(wrapper.state("clusterDomainName")).toBe("Domain")
        expect(wrapper.state("clusterManagedBy")).toBe("Managed By")

    })
})