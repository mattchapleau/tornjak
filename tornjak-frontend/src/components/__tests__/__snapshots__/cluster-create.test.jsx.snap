// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cluster Create Component Should Render 1`] = `
<div>
  <div
    className="cluster-create"
    data-test="cluster-create"
  >
    <div
      className="create-create-title"
      data-test="create-title"
    >
      <h3>
        Create Cluster
      </h3>
    </div>
    <form
      data-test="create-cluster-form"
      onSubmit={[Function]}
    >
      <br />
      <br />
      <div
        className="entry-form"
      >
        <div
          className="clustername-input-field"
          data-test="clustername-input-field"
        >
          <TextInput
            aria-required="true"
            data-test="clustername-Text-input-field"
            helperText="i.e. exampleabc"
            id="clusterNameInputField"
            invalidText="A valid value is required - refer to helper text below"
            labelText="Cluster Name [*required]"
            onChange={[Function]}
            placeholder="Enter CLUSTER NAME"
            required={true}
            role="cluster-name"
          />
        </div>
        <div
          className="clustertype-drop-down"
          data-test="clustertype-drop-down"
        >
          <Dropdown
            aria-required="true"
            ariaLabel="clustertype-drop-down"
            direction="bottom"
            disabled={false}
            helperText=""
            id="clustertype-drop-down"
            itemToElement={null}
            itemToString={[Function]}
            items={Array []}
            label="Select Cluster Type"
            light={false}
            onChange={[Function]}
            role="cluster-type"
            titleText="Cluster Type [*required]"
            type="default"
          />
          <p
            className="cluster-helper"
          >
            i.e. Kubernetes, VMs...
          </p>
        </div>
        <div
          className="cluster-domain-name-input-field"
          data-test="cluster-domain-name-input-field"
        >
          <TextInput
            helperText="i.e. example.org"
            id="clusterDomainNameInputField"
            invalidText="A valid value is required - refer to helper text below"
            labelText="Cluster Domain Name/ URL"
            onChange={[Function]}
            placeholder="Enter CLUSTER DOMAIN NAME/ URL"
          />
        </div>
        <div
          className="cluster-managed-by-input-field"
        >
          <TextInput
            data-test="cluster-domain-name-input-text-field"
            helperText="i.e. person-A"
            id="clusterNameInputField"
            invalidText="A valid value is required - refer to helper text below"
            labelText="Cluster Managed By"
            onChange={[Function]}
            placeholder="Enter CLUSTER MANAGED BY"
          />
        </div>
        <div
          className="agents-multiselect"
          data-test="agents-multiselect"
        >
          <FilterableMultiSelect
            ariaLabel="Choose an item"
            compareItems={[Function]}
            direction="bottom"
            disabled={false}
            filterItems={[Function]}
            helperText="i.e. spiffe://example.org/agent/myagent1..."
            id="selectors-multiselect"
            initialSelectedItems={Array []}
            itemToString={[Function]}
            items={
              Array [
                Object {},
              ]
            }
            label="Select Agents"
            light={false}
            locale="en"
            onChange={[Function]}
            open={false}
            placeholder="Select Agents"
            selectionFeedback="top-after-reopen"
            sortItems={[Function]}
            titleText="Assign Agents To Cluster"
          />
        </div>
        <div
          className="selectors-textArea"
          data-test="selectors-textArea"
        >
          <TextArea
            cols={50}
            defaultValue=""
            disabled={true}
            enableCounter={false}
            helperText="i.e. spiffe://example.org/agent/myagent1..."
            id="selectors-textArea"
            invalid={false}
            invalidText="A valid value is required"
            labelText="Assigned Agents"
            light={false}
            onChange={[Function]}
            onClick={[Function]}
            placeholder="Assigned agents will be populated here - Refer to Assign Agents To Cluster"
            rows={8}
          />
        </div>
        <div
          className="form-group"
          data-test="create-cluster-button"
        >
          <input
            className="btn btn-primary"
            type="submit"
            value="Create Cluster"
          />
        </div>
        <div
          data-test="success-message"
        />
        <div
          className="alert-primary"
          data-test="alert-primary"
          role="alert"
        >
          <pre />
        </div>
      </div>
    </form>
  </div>
  <ToastContainer
    autoClose={5000}
    className="carbon-toast"
    closeButton={[Function]}
    closeOnClick={true}
    containerId="notifications"
    draggable={false}
    draggableDirection="x"
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    role="alert"
    rtl={false}
    theme="light"
    transition={[Function]}
  />
</div>
`;
